### ISSUE TITLE
TypeError on unset `database_key` when using `--hypothesis-seed`

### ISSUE THREAD

**Issue by matthijskooijman** [HypothesisWorks/hypothesis#1254]
I'm running some hypothesis tests under pytest, and was using a fixed
seed using the `--hypothesis-seed` commandline option. At some point, I
got the below exception.

```
    return b'.'.join((self.database_key, b'secondary'))
TypeError: sequence item 0: expected a bytes-like object, NoneType found
```

Apparently the database_key is `None` here. A little investigation shows
that when using a custom seed, the `ConjectureRunner` is called without
a `database_key`:

https://github.com/HypothesisWorks/hypothesis/blob/c877d1d7690c73a793bbaa02c1d9dd62e99aa9d0/hypothesis-python/src/hypothesis/core.py#L712-L715

It also seems that most methods check for a missing `database_key`, but
apparently `downgrade_buffer` does not:

https://github.com/HypothesisWorks/hypothesis/blob/c877d1d7690c73a793bbaa02c1d9dd62e99aa9d0/hypothesis-python/src/hypothesis/internal/conjecture/engine.py#L379-L389

I suspect this is a matter of adding a check, but I'm not too familiar
with the code.

I have made a copy of my test setup to reproduce this bug. If needed, I
can see if I can produce a smaller example triggering this bug.

Here's the full backtrace:

```python-traceback
Traceback (most recent call last):
  File "/home/matthijs/docs/MKIT/3Devo/FeFirmware/tools/test/pytest/test.py", line 166, in test_from_raw
    @given(
  File "hypothesis/core.py", line 1045, in wrapped_test
    state.run()
  File "hypothesis/core.py", line 731, in run
    runner.run()
  File "hypothesis/internal/conjecture/engine.py", line 438, in run
    self._run()
  File "hypothesis/internal/conjecture/engine.py", line 773, in _run
    self.shrink_interesting_examples()
  File "hypothesis/internal/conjecture/engine.py", line 815, in shrink_interesting_examples
    self.shrink(example, predicate)
  File "hypothesis/internal/conjecture/engine.py", line 851, in shrink
    s.shrink()
  File "hypothesis/internal/conjecture/engine.py", line 1276, in shrink
    self.greedy_shrink()
  File "hypothesis/internal/conjecture/engine.py", line 1299, in greedy_shrink
    self.zero_draws()
  File "hypothesis/internal/conjecture/engine.py", line 1404, in zero_draws
    prefix + hbytes(ex.length) + suffix
  File "hypothesis/internal/conjecture/engine.py", line 1249, in cached_test_function
    result = self.__engine.cached_test_function(buffer)
  File "hypothesis/internal/conjecture/engine.py", line 914, in cached_test_function
    self.test_function(result)
  File "hypothesis/internal/conjecture/engine.py", line 243, in test_function
    self.downgrade_buffer(existing.buffer)
  File "hypothesis/internal/conjecture/engine.py", line 389, in downgrade_buffer
    self.database_key, self.secondary_key, buffer)
  File "hypothesis/internal/conjecture/engine.py", line 393, in secondary_key
    return b'.'.join((self.database_key, b'secondary'))
TypeError: sequence item 0: expected a bytes-like object, NoneType found
```

**Comment by Zac-HD** [comment]
Thanks for this excellent bug report!

@DRMacIver, do you know of a reason that adding a check here wouldn't work?

**Comment by DRMacIver** [comment]
> Thanks for this excellent bug report!

Seconded!

> do you know of a reason that adding a check here wouldn't work?

No, that should be fine. I feel like we should probably refactor a bunch of this code to not need this check everywhere, but there's no reason we can't do the simple fix for this bug and then revisit that later.

**Comment by matthijskooijman** [comment]
I already tried the simple fix, which fixes the exception and does not have any obvious side effects (did not do much testing, though):

```
    def downgrade_buffer(self, buffer):                                        
        if self.settings.database is not None:                                 
            if self.database_key is None:                                      
                return
            self.settings.database.move(                                       
                self.database_key, self.secondary_key, buffer)    
```

**Comment by Zac-HD** [comment]
Looks good!  Would you like to make a pull request with this change?  If so, see [our contributing guide](https://github.com/HypothesisWorks/hypothesis/blob/master/CONTRIBUTING.rst) and let me know if there's anything you'd like a hand with; if not one of us can do it :smile:

**Comment by Zac-HD** [comment]
Closed by f0045243da44f4ef51453a14180464e179219c23.

**Comment by matthijskooijman** [comment]
Sorry, never got around to actually making the PR (haven't been working on the related project since then). Thanks for picking this up, change looks good to me :-)
